nova_vm/src/ecmascript/abstract_operations/operations_on_iterator_objects.rs

```diff
@@ -0,0 +1,302 @@
+    let next_method = get(agent, iterator, property)?;
+
+    // 4. Let iteratorRecord be the Iterator Record { [[Iterator]]: iterator, [[NextMethod]]: nextMethod, [[Done]]: false }.
+    // 5. Return iteratorRecord.
+    todo!()
+}
+
+/// [7.4.3 GetIterator ( obj, kind )](https://tc39.es/ecma262/#sec-getiterator)
+///
+/// The abstract operation GetIterator takes arguments obj (an ECMAScript
+/// language value) and kind (sync or async) and returns either a normal
+/// completion containing an Iterator Record or a throw completion.
+pub(crate) fn get_iterator(agent: &mut Agent, obj: Value, is_async: bool) -> JsResult<Object> {
+    // 1. If kind is async, then
+    let method = if is_async {
+        // a. Let method be ? GetMethod(obj, @@asyncIterator).
+        let method = get_method(
+            agent,
+            obj,
+            PropertyKey::Symbol(WellKnownSymbolIndexes::AsyncIterator.into()),
+        )?;
+
+        // b. If method is undefined, then
+        if method.is_none() {
+            // i. Let syncMethod be ? GetMethod(obj, @@iterator).
+            let Some(sync_method) = get_method(
+                agent,
+                obj,
+                PropertyKey::Symbol(WellKnownSymbolIndexes::Iterator.into()),
+            )?
+            else {
+                // ii. If syncMethod is undefined, throw a TypeError exception.
+                return Err(
+                    agent.throw_exception(ExceptionType::TypeError, "No iterator on object")
+                );
+            };
+
+            // iii. Let syncIteratorRecord be ? GetIteratorFromMethod(obj, syncMethod).
+            let sync_iterator_record = get_iterator_from_method(agent, obj, sync_method)?;
+
+            // iv. Return CreateAsyncFromSyncIterator(syncIteratorRecord).
+            todo!("Implement create_async_from_sync_iterator(sync_iterator_record)")
+        } else {
+            method
+        }
+    } else {
+        // 2. Else,
+        // a. Let method be ? GetMethod(obj, @@iterator).
+        get_method(
+            agent,
+            obj,
+            PropertyKey::Symbol(WellKnownSymbolIndexes::Iterator.into()),
+        )?
+    };
+
+    // 4. Return ? GetIteratorFromMethod(obj, method).
+    if let Some(method) = method {
+        get_iterator_from_method(agent, obj, method)
+    } else {
+        // 3. If method is undefined, throw a TypeError exception.
+        Err(agent.throw_exception(
+            ExceptionType::TypeError,
+            "Iterator method cannot be undefined",
+        ))
+    }
+}
+
+/// [7.4.4 IteratorNext ( iteratorRecord [ , value ] )](https://tc39.es/ecma262/#sec-iteratornext)
+///
+/// The abstract operation IteratorNext takes argument iteratorRecord (an
+/// Iterator Record) and optional argument value (an ECMAScript language value)
+/// and returns either a normal completion containing an Object or a throw
+/// completion.
+pub(crate) fn iterator_next(
+    agent: &mut Agent,
+    iterator_record: Object,
```

Comment:
**issue**: As above, the iterator record should be a struct. I think we'd take it by reference here.